codalab/worker/worker.py

```diff
@@ -269,15 +269,20 @@ def start(self):
             self.dependency_manager.start()
         while not self.terminate:
             try:
-                self.process_runs()
-                self.save_state()
                 self.checkin()
-                self.check_termination()
-                self.save_state()
-                if self.check_idle_stop() or self.check_num_runs_stop():
-                    self.terminate = True
-                else:
-                    time.sleep(self.checkin_frequency_seconds)
+                last_checkin = time.time()
+                # Process runs until it's time for the next checkin.
+                while not self.terminate and (
```

Comment:
For this check can we just use threading.timer? In this case it won't block at all.

![image](https://user-images.githubusercontent.com/29891066/155186090-0d95c190-b8bd-442b-a88b-1a98f657bad0.png)

Reply:
I don't think so. If we use threading.Timer, we won't block, that's correct, but then we'll be running `process_runs` only once very 5 seconds. What makes bundles much faster is that we need to run `process_runs` much more often -- not rate-limit it as much as `self.checkin` is rate-limited.